### PR TITLE
Exercise 2

### PR DESCRIPTION
**Pull request by DusanRile**

### CODE REVIEW

File: src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt [hunk 1]
```diff
@@ -18,7 +18,60 @@ import org.jetbrains.exercise2.task3.findPairWithBiggestDifference
  */
 
 internal fun List<Int>.findHighestSumPair(): Pair<Int, Int> {
-    TODO("Implement me!!")
+
+
+//    for ( i in 0  ..this.size - 1){
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt:23]: Please clean up all the comments before pushing the changes and creating the PR.

File: src/main/kotlin/exercise2/task2/FindPairOfHighestSumFunctional.kt [hunk 2]
```diff
@@ -20,7 +20,10 @@ import org.jetbrains.exercise2.task3.findPairWithBiggestDifference
  */
 
 internal fun List<Int>.findHighestSumPairFunctional(): Pair<Int, Int> {
-    TODO("Implement me!!")
+//    var par = Pair<Int, Int>(0, 0)
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task2/FindPairOfHighestSumFunctional.kt:23]: Please clean up all the comments before pushing the changes and creating the PR.

File: src/main/kotlin/exercise2/task3/ConvertMethodToFunctionalApproach.kt [hunk 3]
```diff
@@ -17,24 +17,25 @@ import kotlin.math.abs
...
-
-            if (absDifference >= biggestDifference) {
-                biggestDifference = absDifference
-                resultPair = Pair(first, second)
-            }
-        }
-    }
 
-    return resultPair!!
+//    var resultPair: Pair<Int, Int>? = null
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task3/ConvertMethodToFunctionalApproach.kt:21]: Please clean up all the comments before pushing the changes and creating the PR.

File: src/main/kotlin/exercise2/task4/ProcessCountriesData.kt [hunk 4]
```diff
@@ -59,27 +59,37 @@ internal val countries = listOf(
...
 
 internal fun List<Country>.findCountryWithLowestPopulationDensity(): Country {
-    TODO("Implement me!!!")
+    return this.minBy { it.population/it.totalAreaInSquareKilometers }
 }
 
 internal fun List<Country>.findLanguageSpokenInMostCountries(): String {
-    TODO("Implement me!!!")
+    val jezik = this.flatMap { it.languages }.groupingBy { it }.eachCount().maxBy { it.value}.key
+//    println("Ovo je nesto: " + jezik)
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task4/ProcessCountriesData.kt:79]: Please clean up all the comments before pushing the changes and creating the PR.

File: src/main/kotlin/exercise2/task4/ProcessCountriesData.kt [hunk 5]
```diff
@@ -59,27 +59,37 @@ internal val countries = listOf(
...
+    return jezik
 }
 
 internal fun List<Country>.filterCountriesThatSpeakLanguage(language: String): List<Country> {
-    TODO("Implement me!!!")
+
+//    var lista = listOf<Country>(this.first())
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task4/ProcessCountriesData.kt:85]: Please clean up all the comments before pushing the changes and creating the PR.

File: src/main/kotlin/exercise2/task4/ProcessCountriesData.kt [hunk 6]
```diff
@@ -59,27 +59,37 @@ internal val countries = listOf(
...
 }
 
 internal fun List<Country>.findCountryWithLowestPopulationDensity(): Country {
-    TODO("Implement me!!!")
+    return this.minBy { it.population/it.totalAreaInSquareKilometers }
 }
 
 internal fun List<Country>.findLanguageSpokenInMostCountries(): String {
-    TODO("Implement me!!!")
+    val jezik = this.flatMap { it.languages }.groupingBy { it }.eachCount().maxBy { it.value}.key
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task4/ProcessCountriesData.kt:78]: Please declare variable names in English.

File: src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt [hunk 7]
```diff
@@ -18,7 +18,60 @@ import org.jetbrains.exercise2.task3.findPairWithBiggestDifference
...
+//    for ( i in indices){
+//        if((this[i] > second) && (this[i] < first)){
+//            second = this[i]
+//        }
+//    }
+
+//    Prolazak kroz jednu petlju
+    var first = this[0]
+    var second = this[0]
+    for (i in 0 ..< this.size -1){
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt:65]: Nicely done with a single loop! 👏🏻 

nit: `i in 0 ..< this.size -1` can be replaced with `i in this.indices`

File: src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt [hunk 8]
```diff
@@ -18,7 +18,60 @@ import org.jetbrains.exercise2.task3.findPairWithBiggestDifference
...
+        if (this[i+1] >= first ){
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt:66]: We are accessing `this[i + 1]` many times, hence we assign this to variable `nextNum` and use `nextNum` variable everywhere else. This will improve readability.

File: src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt [hunk 9]
```diff
@@ -18,7 +18,60 @@ import org.jetbrains.exercise2.task3.findPairWithBiggestDifference
...
+            second = first
+            first = this[i+1]
+        }
+        else if ((second < this[i+1]) && second < first){
+            second = this[i+1]
+        }
+    }
+    return Pair<Int, Int>(first, second)
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt:74]: The `<Int, Int>` can be omitted since the compiler is smart enough to infer the generic type.

Also, there is an alternative way to declare a pair, and that is by using `to` infix extension function:
```first to second```

File: src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt (hunk 7 above)
Comment (vuksa) [inline src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt:65]: If we run your implementation for the input `listOf(3, 1, 2)`, we would get the incorrect result `Pair(3, 3)`. 
Are the tests green for you?

nit: `this.size -1` can be written as `this.lastIndex`

Comment (vuksa) [reply src/main/kotlin/exercise2/task1/FindPairOfHighestSum.kt:65]: The test data for this method seems incomplete, and your implementation has a false positive in the tests, meaning the tests are green, but the actual implementation is incorrect.  

In the above example for the input `listOf(3, 1, 2)` the expected output should be `Pair(3, 2)`, and you implementation returns `Pair(3, 3)` which is incorrect. This is because the test didn't cover the case where the highest number in the list is at the first index. 

I'm very sorry for the inconvenience with incomplete test data.

Please add a missing case to the test class and fix the bug in your implementation.

File: src/main/kotlin/exercise2/task3/ConvertMethodToFunctionalApproach.kt [hunk 10]
```diff
@@ -17,24 +17,25 @@ import kotlin.math.abs
...
+//
+//            if (absDifference >= biggestDifference) {
+//                biggestDifference = absDifference
+//                resultPair = Pair(first, second)
+//            }
+//        }
+//    }
+//    return resultPair!!
+
+    return this.sorted().let { it.first() to it[lastIndex] }
```
Comment (vuksa) [inline src/main/kotlin/exercise2/task3/ConvertMethodToFunctionalApproach.kt:38]: There is a mistake in the implementation where the `lastIndex` is stated. In the code you wrote, the `lastIndex` is the last index of the receiver object list. It's the same as if we wrote `this.lastIndex`.

TLDR: you should use `it.lastIndex` instead of `this.lastIndex`

In this case, the result will be correct, but if we, for example, filter elements of the list after sorting it, the `lastIndex` of the list we are operating on would change, and you would get an `IndexOutOfBounds` error. 

Let's show this visually. In the following example, we have the list `[3, 2, 4, 5, 1]`.
After calling `sorted()` on the list, we would get a new list `[1, 2, 3, 4, 5]`, with the last index value being `4`.
If after that, on the result list we call `filter { it > 2 }`, the new result list would be `[3, 4, 5]`, where now the value of the last index is `2`, and if we would call `it[this.lastIndex]` as in your code, we would get `IndexOutOfBoundsException` because `this.lastIndex` value is 4, which is greater than current last index.
